freqtrade/freqai/freqai_interface.py

```diff
@@ -526,14 +527,24 @@ def model_exists(self, dk: FreqaiDataKitchen) -> bool:
         return file_exists
 
     def set_full_path(self) -> None:
+        """
+        Creates and sets the full path for the identifier
+        """
         self.full_path = Path(
-            self.config["user_data_dir"] / "models" / f"{self.freqai_info['identifier']}"
+            self.config["user_data_dir"] / "models" / f"{self.identifier}"
         )
         self.full_path.mkdir(parents=True, exist_ok=True)
-        shutil.copy(
-            self.config["config_files"][0],
-            Path(self.full_path, Path(self.config["config_files"][0]).name),
-        )
+
+    def record_configs(self) -> None:
+        """
+        Records configs in the full path for reproducibility
+        """
+        self.config_record_path = self.full_path / Path("configs")
+        self.config_record_path.mkdir(parents=True, exist_ok=True)
+
+        for config in self.config["config_files"]:
+            dest_config_path = self.config_record_path / Path(config).name
+            shutil.copyfile(config, dest_config_path)
```

Comment:
I'm a bit concerned about this (not this PR per se, but the whole functionality of "copy config someplace else").
If we assume a live configuration, we're effectively copy/pasting people's exchange keys to random places (within the user directory).
People may take extra precautions (permissions, ..., for example) for their keys - which we now basically void by copying the config to other places.

@robcaulk can you remind me of the purpose of doing a copy of the config files? (why do we need a copy of the config file in the model directory) ?

Reply:
You make a good point. 

The reason we copy the config over is because it is a record of the run. This enables improved reproducibility, ease of parameter exploration (easy to see which parameters were associated with a set of results).

we could parse it to save everything except the sensitive key information?

Reply:
Would it make more sense to only save certain information such as the Freqai config, timeframe, etc?

Reply:
if reproducibiliy is the goal, then i think copying json files is the wrong approach - neglecting most possibilities freqtrade offers to configure things (you're also omitting the strategy - which is also a big part of reproducing (or not) a result.

i can see several ways where the above simply won't work
*  `-c config.json -c secret/config.json -c ../someother/config.json` (both names are identical ...) - this could go on with 20 directories
* `export FREQTRADE__FREQAI__TRAIN_PERIOD_DAYS=20` (while the config defines this as 15) - so your "reproducibility" would show a value of 15 - but the bot runs with 20 (obviously this is possible for every value)
* `--timeframe 1m` (config / strategy defines the timeframe as 5m - but the runtime argument overwrites this
* config not defining timeframe, but the strategy defines the timeframe (that's how i run all my bots) - so you'd not capture the actual timeframe used (this is obviously true for many parameters - but timeframe is the most relevant one i think

the "running" config (basically the `config` object we pass around throughout the bot) will contain the above info - compiled in the right sequence - so that's the only thing that REALLY defines what's running.

Now simply serializing this will be difficult - as it'll contain 
* sensitive information
* non-serializable objects like paths

While we can identify the first pretty easily - i think it's better to pick what we really need (what's actually relevant to reproducibility). That'll avoid potentially leaking new sensitive information (new keys) - and will be simpler than to identify the non-json objects we keep in config (like all the Path objects).

Reply:
I understand Freqtrade is perhaps too flexible on that front of parameter setting. I’m not aiming to solve that “issue” anytime soon. I think you are a bit gun shy here. You forgot to point out that the entire source code version is also changeable underneath ;). 

The point is to just have a record of parameters stored with the results/models. This includes pairlist, pairlist filter, stake size etc.

Sure copy the strat over too, that’s a good idea. I had it in earlier versions of FreqAI, im happy to bring it back.

The solution here is to simply copy the full config without any secret-containing sub dictionaries. Am I missing something?

Reply:
> I think you are a bit gun shy here

I'm not sure i understand your point (especially, i don't understand the point of getting personal)

I'm pointing out flaws in the current logic - which is NOT fulfilling your need of having "a record of parameters".
The new logic (copy all additional configs) will make it worse - as you might end up overwriting the most important config in case of naming collisions.

It may work for you personally by copying the configuration - but that's not the case for most users (it's rare to see a configuration containing timeframe - that's usually coming from the strategy).
That's different with the freqAI examples, but they're really the exception - the recommendation is to have the timeframe in the strategy - not in the configuration.

> The solution here is to simply copy the full config without any secret-containing sub dictionaries. Am I missing something?

If that's the goal - then the current behavior is correct - and the behavior (what this PR is introducing) would revert this by including all files.

Reply:
It seems the best solution is to simply wait for full initialization of the `IFreqaiModel` object, and then grab the `freqai` relevant parameters from the configuration dictionary:

`freqai, timeframe, stake_amount, max_open_trades, pairs`

Now we no longer need to save any particular config file, and we avoid copying secrets unnecessarily.

Reply:
I'm not sure what you mean by the current way is correct, as it just grabs the first config file and copies that. Which could be anything, the full config, your telegram config file, etc.

Reply:
> I'm not sure what you mean by the current way is correct

That's a direct answer to the sentence quoted - which says "The solution here is to simply copy the full config without any secret-containing sub dictionaries.". 
if this would be the aim (it isn't!) - then current behavior would be correct (it's based on the assumption that secrets are in subdirectories - which is not the case).

> wait for full initialization of the `IFreqaiModel` object

do we need to wait for this?
is freqAI changing settings in the configuration if it determines some invalid state, or is it just warning / failing on these conditions?

Reply:
IFreqaiModel has some automatic "reorientation" to help ensure users aren't getting themselves into trouble. So it needs to wait until initialization finishes on the IFreqaiModel object.